lib/auth/server_monitor.go

```diff
@@ -0,0 +1,148 @@
+)
+
+// MonitorSystemTime runs the periodic check for iterating through all inventories
+// to ping them and receive the system clock difference.
+func (a *Server) MonitorSystemTime(ctx context.Context) error {
+	checkInterval := interval.New(interval.Config{
+		FirstDuration: time.Minute,
+		Duration:      systemClockCheckCycle,
+		Clock:         a.GetClock(),
+		Jitter:        retryutils.NewSeventhJitter(),
+	})
+	defer checkInterval.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-checkInterval.Next():
+			var inventories []string
+			a.inventory.Iter(func(handle inventory.UpstreamHandle) {
+				info := handle.Hello()
+				diff, err := handle.TimeReconciliation(ctx, insecurerand.Uint64())
+				if err != nil {
+					slog.ErrorContext(ctx, "error getting time reconciliation")
+					return
+				}
+
+				if (diff > 0 && diff > systemClockThreshold) || (diff < 0 && -diff > systemClockThreshold) {
+					inventories = append(inventories, fmt.Sprintf(
+						" - %s[%s] is %s",
+						info.GetServerID(),
+						types.SystemRoles(info.GetServices()).String(),
+						durationText(diff),
+					))
+					slog.WarnContext(ctx, "server time difference detected",
+						"server", info.GetServerID(),
+						"services", info.GetServices(),
+						"difference", durationText(diff),
+					)
+				}
+			})
+
+			if len(inventories) > 0 {
+				err := upsertGlobalNotification(ctx, a.Services, generateNotificationMessage(inventories))
+				if err != nil {
+					slog.ErrorContext(ctx, "can't set notification about system clock issue", "error", err)
+					continue
+				}
+			}
+		}
+	}
+}
+
+// upsertGlobalNotification sets predefined global notification for notifying the issues with the cluster
+// servers related to the system clock difference in nodes.
+func upsertGlobalNotification(ctx context.Context, services *Services, text string) error {
+	_, err := services.UpsertGlobalNotification(ctx, &notificationsv1.GlobalNotification{
+		Kind:     types.KindGlobalNotification,
+		Version:  types.V1,
+		Metadata: &headerv1.Metadata{Name: systemClockNotificationWarningName},
+		Spec: &notificationsv1.GlobalNotificationSpec{
+			Matcher: &notificationsv1.GlobalNotificationSpec_All{
+				All: true,
+			},
+			Notification: &notificationsv1.Notification{
+				SubKind: types.NotificationDefaultWarningSubKind,
+				Spec:    &notificationsv1.NotificationSpec{},
+				Metadata: &headerv1.Metadata{
+					Name: systemClockNotificationWarningName,
+					Labels: map[string]string{
+						types.NotificationTitleLabel: text,
+					},
+				},
+			},
+		},
+	})
+	return trace.Wrap(err)
+}
+
+// generateNotificationMessage formats the notification message with the inventory list.
+func generateNotificationMessage(messages []string) string {
+	var more string
+	if len(messages) > 10 {
+		more = fmt.Sprintf("Only 10 servers are shown out of the list of %d.", len(messages))
+		messages = messages[:10]
+	}
```

Comment:
inventory iterator doesn't support limit, so we have to truncate or add the logic to inventory store

Reply:
Does it really make sense to monitor servers periodically? Are we actually concerned about _drift_ rather than just misconfigurations?

And if we are, why an "external" periodic process rather than something that runs as part of the handling of the inventory control stream?

Reply:
Generally periodically it runs to gather info about all nodes in cluster to generate one report with the list of nodes. I believe in most cases original issue related to misconfiguration, but there are other cases as well like this one with VM https://github.com/gravitational/teleport/issues/22166#issuecomment-2042862626 , another might be with the node in private network without local ntp server (not sure how much time needs for RTC to reach this 1min threshold inaccuracy)

@espadolini do you mean it should be a part of the inventory control stream (like periodic check and generating the list of nodes with time inconsistency and api to fetch this list)?

Reply:
If this was part of the control stream handling you could just have the measured drift (if available) as part of the `upstreamHandle` and potentially added to the instance heartbeats, and any periodic report operation would just iterate the store and copy off the data, and then decide what to do about the report.